### MISY408 - Project 6

### Titanic Dataset

**Attributes**

* Survived: 0 = Not survived, 1 = Survived
* Pclass: Ticket class 1 = 1st class, 2 = 2nd class, 3 = 3rd class
* Name: Name of passenger
* Sex: male, female
* Age: Age of passenger
* Sibsp: # of siblings / spouses aboard the Titanic
* Parch: # of parents / children aboard the Titanic
* Ticket: Ticket number
* Fare: Passenger fare
* Cabin: Cabin number
* Embarked: Port of Embarkation C = Cherbourg, Q = Queenstown, S = Southampton

**Question1: Import the required libraries and read Titanic.csv file and save it in df dataframe, display the first five rows of df dataframe**

In [1]:
import pandas as pd
df = pd.read_csv('Titanic.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

**Question 2: Exploratory data analysis. For this part you should:**

        1) check the shape of dataset
        2) View the statistical summary of numerical variables (describe)

In [2]:
df.shape

(891, 12)

In [3]:
df.describe().transpose()

count        mean         std   min       25%       50%    75%  \
PassengerId  891.0  446.000000  257.353842  1.00  223.5000  446.0000  668.5   
Survived     891.0    0.383838    0.486592  0.00    0.0000    0.0000    1.0   
Pclass       891.0    2.308642    0.836071  1.00    2.0000    3.0000    3.0   
Age          714.0   29.699118   14.526497  0.42   20.1250   28.0000   38.0   
SibSp        891.0    0.523008    1.102743  0.00    0.0000    0.0000    1.0   
Parch        891.0    0.381594    0.806057  0.00    0.0000    0.0000    0.0   
Fare         891.0   32.204208   49.693429  0.00    7.9104   14.4542   31.0   

                  max  
PassengerId  891.0000  
Survived       1.0000  
Pclass         3.0000  
Age           80.0000  
SibSp          8.0000  
Parch          6.0000  
Fare         512.3292

**Question 3:** 

a. Create a frequency table to show the number of survived and number of not-survived passengers

In [4]:
df.Survived.value_counts()

0    549
1    342
Name: Survived, dtype: int64

b. Create a crosstab to show the number of survived and the number of non-survived passengers for each Pclass (interpret the results. for example: which Pclass's passengers survived the most?)

In [5]:
pd.crosstab(df['Survived'], df['Pclass'])

Pclass      1   2    3
Survived              
0          80  97  372
1         136  87  119

## According to this crosstab: 136 passangers from class one survived,
## 87 from class two survived, and 119 from class 3 survived. 

## 80 passangers from class 1 did not survive, 97 from class 2 did not survive and 372 from did not survive


c. Create a crosstab to show the number of survived and the number of not-survived passengers for each Sex (interpret the results. for example: which Sex's passengers survived the most?)

In [6]:
pd.crosstab(df['Survived'], df['Sex'])

Sex       female  male
Survived              
0             81   468
1            233   109

## According to this crostab: more female passangers survived compared to male, to be exact 233 females survivied and 109 males survived. Also more male passangers did not survive compared to female passangers: 81 female, and 468 male did not survive. 

**Question 4:**

a. Set X variable to 'Sex' and 'Plass' and set y variable to 'Survived'.

(Note: Since Sex is categorical variable you should code it. For example: create SexCode and set male to 0 and female to 1)

In [7]:
df.loc[df['Sex']=='male', 'Sex'] = 0
df.loc[df['Sex']=='female', 'Sex'] = 1
y = df['Survived']
X= df[['Pclass','Sex']]
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris   0  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...   1  38.0      1   
2                               Heikkinen, Miss. Laina   1  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)   1  35.0      1   
4                             Allen, Mr. William Henry   0  35.0      0   
..                                                 ...  ..   ...    ...   
886                              Montvila, Rev. Juozas   0  27.0      0   
887                       Graham, Miss. Margaret Edith   1  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"   1   NaN      1   
889                              Behr, Mr. Karl Howell   0  26.0      0   
890                                Dooley, Mr. Patrick   0  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

b. Create a Decision Tree

In [8]:
from sklearn import tree
clf= tree.DecisionTreeClassifier(max_depth=3, random_state = 0)
clf= clf.fit(X, y)

from IPython.display import Image  
from six import StringIO  
import pydotplus

dot_data = StringIO()  
tree.export_graphviz(clf, out_file=dot_data,  feature_names=X.columns, rounded=True, rotate=True)
graph = pydotplus.graph_from_dot_data(dot_data.getvalue())  
Image(graph.create_png())  



InvocationException: GraphViz's executables not found

d. Based on Gini score which node has the best classification (has lower impurity)? which node had the worst classification (has higher impurity)?

The lowest impurity is in the 4th colmn with a gini score of .062

the highest impurity is in the 3rd colmn with the gini score of .5

**Question 5:**

a. In the following, type a code to show the importance of each feature. 

In [11]:
clf.feature_importances_

array([0.24987426, 0.75012574])

b. In the following, Interpret the feature importance output

The array above tells us that the sex of a passanger matters more than what class, the passanger class they accounted for. Therefore rougly 25% accounted for importance and 75% accounted for the sex. 

**Question6:**

a. In the following, type a code to create a confusion matrix

In [9]:
predictions= clf.predict(X)
from sklearn.metrics import confusion_matrix
# suggestion: put ACTUAL value first, predictions second.
cm = confusion_matrix(y, predictions)
cm


array([[540,   9],
       [181, 161]], dtype=int64)

b. In the following, Interpret the confusion matrix



True Negative : Of the people who did not survive the titanic, we predicated 540 correctly. 
True Positive: of the people survived the titanic, we predicated 161 correctly. 

False Positive: Of the people who did not survive the titanic , we predicated 9 to survived(incorrectly)

false negative: of the people who did survive the titanic, we predicated 181 to not survive.

**Question 7: In the following, type a code to show the accuracy of your model?**

In [10]:
accuracy = clf.score(X,y) 
print("Your decision tree is {:.2f}% accurate.".format(accuracy*100))

Your decision tree is 78.68% accurate.
